pkg/controller/common/helpers.go

```diff
@@ -250,30 +252,145 @@ func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
 	}
 
 	switch typedConfig := ignconfigi.(type) {
+	case ign3types.Config:
+		return ignconfigi.(ign3types.Config), nil
 	case ign2types.Config:
-		return ignconfigi.(ign2types.Config), nil
+		convertedIgnV3, err := ConvertIgnition2to3(ignconfigi.(ign2types.Config))
+		if err != nil {
+			return ign3types.Config{}, errors.Wrapf(err, "failed to convert Ignition config spec v2 to v3")
+		}
+		return convertedIgnV3, nil
+	default:
+		return ign3types.Config{}, errors.Errorf("unexpected type for ignition config: %v", typedConfig)
+	}
+}
+
+// Test: have some situations de-duplicate and see what happens, will merge into main function
+func ParseAndConvertConfigWithDeduplication(rawIgn []byte) (ign3types.Config, error) {
+	ignconfigi, err := IgnParseWrapper(rawIgn)
+	if err != nil {
+		return ign3types.Config{}, errors.Wrapf(err, "failed to parse Ignition config")
+	}
+
+	switch typedConfig := ignconfigi.(type) {
 	case ign3types.Config:
-		convertedIgnV2, err := ConvertIgnition3to2(ignconfigi.(ign3types.Config))
+		return ignconfigi.(ign3types.Config), nil
+	case ign2types.Config:
+		ignconfv2 := removeIgnDuplicateFilesAndUnits(ignconfigi.(ign2types.Config))
+		convertedIgnV3, err := ConvertIgnition2to3(ignconfv2)
 		if err != nil {
-			return ign2types.Config{}, errors.Wrapf(err, "failed to convert Ignition config spec v3 to v2")
+			return ign3types.Config{}, errors.Wrapf(err, "failed to convert Ignition config spec v2 to v3")
 		}
-		return convertedIgnV2, nil
+		return convertedIgnV3, nil
 	default:
-		return ign2types.Config{}, errors.Errorf("unexpected type for ignition config: %v", typedConfig)
+		return ign3types.Config{}, errors.Errorf("unexpected type for ignition config: %v", typedConfig)
 	}
 }
 
+// Function to remove duplicated files/units from a V2 MC, since the translator
+// (and ignition spec V3) does not allow for duplicated entries in one MC.
+// This should really not change the actual final behaviour, since it keeps
+// ordering into consideration and has contents from the highest alphanumeric
+// MC's final version of a file.
+// Note:
+// Append is not considered since we do not allow for appending
+// Units have one exception: dropins are concat'ed
+
+func removeIgnDuplicateFilesAndUnits(ignConfig ign2types.Config) ign2types.Config {
+
+	files := ignConfig.Storage.Files
+	units := ignConfig.Systemd.Units
+
+	filePathMap := map[string]bool{}
+	var outFiles []ign2types.File
+	for i := len(files) - 1; i >= 0; i-- {
+		// We do not actually support to other filesystems so we make the assumption that there is only 1 here
+		path := files[i].Path
+		if duplicate, isDup := filePathMap[path]; isDup {
+			glog.Infof("Found duplicate file: %v, older file will not be rendered.", duplicate)
+			continue
+		}
+		outFiles = append(outFiles, files[i])
+		filePathMap[path] = true
+	}
+
+	// TODO jerzhang check if you can have contents and dropins in the same file
+	unitNameMap := map[string]bool{}
+	var outUnits []ign2types.Unit
```

Comment:
so units is somewhat wonky since it needs to consider dropins when doing a merge, now consider this scenario:
```
Unit A:
  Content: AC
  Dropins:
    Dropin: AD
```
```
Unit B:
    Content: BC
  Dropins:
    Dropin: BD
```
Should the final look like:
```
  Unit rendered:
    Content: BC
  Dropins:
    Dropin: AD
    Dropin: BD
```
or 
```
Unit rendered:
    Content: BC
  Dropins:
    Dropin: BD
```

Reply:
I'd be in favor of the former, having both drop ins added.

Another case that may also be considered is if there is a unit in one template, and a dropin only in another template.

Reply:
In the case of a unit in one and a dropin in another, both the existing spec 2 behaviour and the ignition spec 3 merge function keeps both, so I think we should keep that.

And agreed, having both dropins probably makes it more consistent